examples/vae.py

```diff
@@ -0,0 +1,100 @@
+from __future__ import absolute_import, division, print_function
+
+import argparse
+from collections import OrderedDict
+
+import torch
+import torch.utils.data
+from torch import nn, optim
+from torch.nn import functional as F
+from torchvision import datasets, transforms
+
+import funsor
+import funsor.distributions as dist
+import funsor.ops as ops
+from funsor.domains import bint, reals
+
+
+class Encoder(nn.Module):
+    def __init__(self):
+        super(Encoder, self).__init__()
+        self.fc1 = nn.Linear(784, 400)
+        self.fc21 = nn.Linear(400, 20)
+        self.fc22 = nn.Linear(400, 20)
+
+    def forward(self, image):
+        h1 = F.relu(self.fc1(image))
+        loc = self.fc21(h1)
+        scale = self.fc22(h1)
+        return loc, scale
+
+
+class Decoder(nn.Module):
+    def __init__(self):
+        super(Encoder, self).__init__()
+        self.fc3 = nn.Linear(20, 400)
+        self.fc4 = nn.Linear(400, 784)
+
+    def forward(self, z):
+        h3 = F.relu(self.fc3(z))
+        return torch.sigmoid(self.fc4(h3))
+
+
+def main(args):
+    encoder = Encoder()
+    decoder = Decoder()
+
+    encode = funsor.function(reals(28, 28), (reals(20), reals(20)))(encoder)
+    decode = funsor.function(reals(20), reals(28, 28))(decoder)
+
+    @funsor.interpreter.interpretation(funsor.terms.monte_carlo)
+    def loss_function(data, scale):
+        loc, scale = encode(data)
+        i = funsor.Variable('i', bint(20))
+        z = funsor.Variable('z', reals(20))
+        q = dist.Normal(loc[i], scale[i], value=z[i])
+        q = q.reduce(ops.add, frozenset(['i']))
+
+        probs = decode(z)
+        x = funsor.Variable('x', bint(28))
+        y = funsor.Variable('y', bint(28))
+        p = dist.Bernoulli(probs[x, y], value=data[x, y])
+        p = p.reduce(ops.add, frozenset(['x', 'y']))
+
+        elbo = funsor.Integrate(q, scale * (p - q), 'z')
+        elbo = elbo.reduce(ops.add, frozenset(['batch']))
```

Comment:
I see that the `batch` dim is only being referenced here the first time. Do all the funsors above, like `encode`, `dist.Normal` work with arbitrary batch dims on the left?

Reply:
Yes, we've taken care that all funsors "broadcast" to allow arbitrary batch dims.